spirv_msl.cpp

```diff
@@ -18403,6 +18868,189 @@ void CompilerMSL::emit_block_hints(const SPIRBlock &)
+		}
+		statement("spvPerPrimitive spvP = {};");
+		for (uint32_t index = 0; index < uint32_t(type_prim.member_types.size()); ++index)
+		{
+			uint32_t orig_var =
+			    get_extended_member_decoration(type_prim.self, index, SPIRVCrossDecorationInterfaceOrigID);
+			uint32_t orig_id =
+			    get_extended_member_decoration(type_prim.self, index, SPIRVCrossDecorationInterfaceMemberIndex);
+			auto &orig = get<SPIRVariable>(orig_var);
+			auto &orig_type = get<SPIRType>(orig.basetype);
+
+			BuiltIn builtin = BuiltInMax;
+			std::string access;
+			if (orig_type.basetype == SPIRType::Struct)
+			{
+				if (has_member_decoration(orig_type.self, orig_id, DecorationBuiltIn))
+					builtin = BuiltIn(get_member_decoration(orig_type.self, orig_id, DecorationBuiltIn));
+
+				switch (builtin)
+				{
+				case BuiltInPrimitiveId:
+				case BuiltInLayer:
+				case BuiltInViewportIndex:
+				case BuiltInCullPrimitiveEXT:
+				case BuiltInPrimitiveShadingRateKHR:
+					access = "." + builtin_to_glsl(builtin, StorageClassOutput);
+					break;
+				default:
+					access = "." + to_member_name(orig_type, orig_id);
+				}
+			}
+			statement("spvP.", to_member_name(type_prim, index), " = ", to_name(orig_var), "[spvI]", access, ";");
+		}
+		statement("spvMesh.set_primitive(spvI, spvP);");
+		end_scope();
+	}
+}
+
+void CompilerMSL::emit_mesh_entry_point()
+{
+	auto &ep = get_entry_point();
+	auto &f = get<SPIRFunction>(ir.default_entry_point);
+
+	const uint32_t func_id = ir.increase_bound_by(3);
+	const uint32_t block_id = func_id + 1;
+	const uint32_t ret_id = func_id + 2;
+	current_function = &set<SPIRFunction>(func_id, f.return_type, f.function_type);
+
+	current_function->blocks.push_back(block_id);
+	current_function->entry_block = block_id;
+
+	current_block = &set<SPIRBlock>(block_id);
+	current_block->terminator = SPIRBlock::Return;
+
+	// push call to original 'main'
+	Instruction ix = {};
+	ix.op = OpFunctionCall;
+	ix.offset = uint32_t(ir.spirv.size());
+	ix.length = 3;
+
+	ir.spirv.push_back(f.return_type);
+	ir.spirv.push_back(ret_id);
+	ir.spirv.push_back(ep.self);
+
+	current_block->ops.push_back(ix);
+
+	current_block = nullptr;
+	current_function = nullptr;
+
+	// relace entry-point for new one
+	SPIREntryPoint proxy_ep = ep;
+	proxy_ep.self = func_id;
+	ir.entry_points.insert(std::make_pair(func_id, proxy_ep));
+	ir.meta[func_id] = ir.meta[ir.default_entry_point];
+	ir.meta[ir.default_entry_point].decoration.alias.clear();
+
+	ir.default_entry_point = func_id;
+}
+
+void CompilerMSL::emit_mesh_tasks(SPIRBlock &block)
+{
+	// GLSL: Once this instruction is called, the workgroup must be terminated immediately, and the mesh shaders are launched.
+	// TODO: find relieble and clean of terminating shader.
+	statement("spvMpg.set_threadgroups_per_grid(uint3(", to_unpacked_expression(block.mesh.groups[0]), ", ",
+	          to_unpacked_expression(block.mesh.groups[1]), ", ", to_unpacked_expression(block.mesh.groups[2]), "));");
```

Comment:
`return;` should be called after this. Also needs to check that this is only used in the entry function, otherwise just throw an error saying it's not implemented.

Reply:
I use to have something like this on previous iterations. Having non-trivial behavior make it even worse:
In this stage we can document `shader termination in Emit doesn't work`. 
Otherwise  it will be `shader termination in Emit sometimes work, but may fail to compile if <...>`